pkg/k8s/watchers/cilium_node.go

```diff
@@ -5,167 +5,153 @@ package watchers
-					if ciliumNode == nil {
-						return
-					}
-					n := nodeTypes.ParseCiliumNode(ciliumNode)
-					k.nodeDiscoverManager.NodeDeleted(n)
-				},
-			},
+		var synced atomic.Bool
+		stop := make(chan struct{})
+
+		k.blockWaitGroupToSyncResources(
+			stop,
 			nil,
+			func() bool { return synced.Load() },
+			apiGroup,
 		)
-		isConnected := make(chan struct{})
-		// once isConnected is closed, it will stop waiting on caches to be
-		// synchronized.
-		k.blockWaitGroupToSyncResources(isConnected, swgNodes, ciliumNodeInformer.HasSynced, apiGroup)
-
-		k.ciliumNodeStoreMU.Lock()
-		k.ciliumNodeStore = ciliumNodeStore
-		k.ciliumNodeStoreMU.Unlock()
-
-		once.Do(func() {
-			// Signalize that we have put node controller in the wait group
-			// to sync resources.
-			asyncControllers.Done()
-		})
 		k.k8sAPIGroups.AddAPI(apiGroup)
-		go ciliumNodeInformer.Run(isConnected)
 
-		<-kvstore.Connected()
-		log.Info("Connected to key-value store, stopping CiliumNode watcher")
+		// Signalize that we have put node controller in the wait group to sync resources.
+		once.Do(asyncControllers.Done)
 
-		// Set the ciliumNodeStore as nil so that any attempts of getting the
-		// CiliumNode are performed with a request sent to kube-apiserver
-		// directly instead of relying on an outdated version of the CiliumNode
-		// in this cache.
-		k.ciliumNodeStoreMU.Lock()
-		k.ciliumNodeStore = nil
-		k.ciliumNodeStoreMU.Unlock()
+		// Set forceCiliumNodeFromApiServer to false so that any attempts of getting the
+		// CiliumNode go through the Resource[T] cache.
+		k.forceCiliumNodeFromApiServer.Store(false)
 
-		close(isConnected)
+		// derive another context to signal Events() in case of kvstore connection
+		eventsCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
 
-		k.cancelWaitGroupToSyncResources(apiGroup)
-		k.k8sAPIGroups.RemoveAPI(apiGroup)
-		// Create a new node controller when we are disconnected with the
-		// kvstore
-		<-kvstore.Client().Disconnected()
+		go func() {
+			defer close(stop)
 
-		log.Info("Disconnected from key-value store, restarting CiliumNode watcher")
+			events := k.resources.CiliumNode.Events(eventsCtx)
+			cache := make(map[resource.Key]*cilium_v2.CiliumNode)
+			for event := range events {
+				var err error
+				switch event.Kind {
+				case resource.Sync:
+					synced.Store(true)
+				case resource.Upsert:
+					var needUpdate bool
+					oldObj, ok := cache[event.Key]
+					if !ok {
+						needUpdate = k.onCiliumNodeInsert(event.Object)
+					} else {
+						needUpdate = k.onCiliumNodeUpdate(oldObj, event.Object)
+					}
+					if needUpdate {
+						cache[event.Key] = event.Object.DeepCopy()
+					}
+				case resource.Delete:
+					k.onCiliumNodeDelete(event.Object)
+					delete(cache, event.Key)
+				}
+				event.Done(err)
+			}
+		}()
+
+		select {
+		case <-kvstore.Connected():
+			log.Info("Connected to key-value store, stopping CiliumNode watcher")
```

Comment:
Would there be any reason to wait on the stop channel to close notifying that the event loop has processed all events?

Reply:
I cannot think of any, since AFAIU the current implementation of Resource[T] closes the channel returned by `Events` from the producer side only when the resource is shut down. But in that case, the entire hive is shutting down, so we're gonna end up in the `case <-ctx.Done():`.